### PR TITLE
[Data] Add `partitioning` parameter to `read_parquet`

### PR DESCRIPTION
**Pull request by bveeramani**
## Why are these changes needed?

To extract path partition information with `read_parquet`, you pass a PyArrow `partitioning` object to `dataset_kwargs`. For example:
```
schema = pa.schema([("one", pa.int32()), ("two", pa.string())])
partitioning = pa.dataset.partitioning(schema, flavor="hive")
ds = ray.data.read_parquet(... dataset_kwargs=dict(partitioning=partitioning))
```

This is problematic for two reasons:
1. It tightly couples the interface with the implementation; partitioning only works if we use `pyarrow.Dataset` in a specific way in the implementation.
2. It's inconsistent with all of the other file-based API. All other APIs use expose a top-level `partitioning` parameter (rather than `dataset_kwargs`) where you pass a Ray Data `Partitioning` object (rather than a PyArrow partitioning object). 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by raulchen — APPROVED** [review]

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 1]
```diff
@@ -633,3 +650,17 @@ def sample_fragments(
     sample_bar.close()
 
     return sample_infos
+
+
+def _add_partitions_to_table(table, partitions):
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/datasource/parquet_datasource.py:655]: Please annotate types

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 2]
```diff
@@ -633,3 +650,17 @@ def sample_fragments(
...
+    import pyarrow as pa
+
+    for field, value in partitions.items():
+        column = pa.array([value] * len(table))
+        field_index = table.schema.get_field_index(field)
+        if field_index != -1:
+            table = table.set_column(field_index, field, column)
+        else:
+            table = table.append_column(field, column)
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/datasource/parquet_datasource.py:664]: Same problem that i've called out in [here](https://github.com/ray-project/ray/pull/47551/files#r1753122517) -- partition schema needs to be inferred once for the whole table, not per partition

File: python/ray/data/_internal/datasource/parquet_datasource.py (hunk 1 above)
Comment (alexeykudinkin) [reply python/ray/data/_internal/datasource/parquet_datasource.py:655]: Not addressed

Comment (bveeramani) [reply python/ray/data/_internal/datasource/parquet_datasource.py]: Fixed in 4f15d25c99cae192fe95a37cf196b9253ceb47ed

File: python/ray/data/_internal/datasource/parquet_datasource.py (hunk 2 above)
Comment (bveeramani) [reply python/ray/data/_internal/datasource/parquet_datasource.py]: Discussed offline -- we'll default to string if the type isn't explicitly specified by the user

**Pull request merged**